File: MPP/javafx/src/main/java/org/example/repository/ConcertDatabaseRepositoryImpl.java
package org.example.repository;

import org.example.domain.Concert;

import java.sql.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ConcertDatabaseRepositoryImpl extends AbstractDatabaseRepository<Long, Concert> implements ConcertRepository {

    public ConcertDatabaseRepositoryImpl(Properties properties) {
        super(properties);
    }

    @Override
    protected Concert extractEntity(ResultSet resultSet) throws SQLException {
        return new Concert(
                resultSet.getLong("id"),
                resultSet.getString("artist"),
                resultSet.getTimestamp("date").toLocalDateTime(),
                resultSet.getString("location"),
                resultSet.getInt("available_tickets")
        );
    }

    @Override
    protected String getFindOneQuery() {
        return "select * from concerts where id = ?";
    }

    @Override
    protected void setIdParameter(PreparedStatement statement, Long aLong) throws SQLException {
        statement.setLong(1, Math.toIntExact(aLong));
    }

    @Override
    protected String getFindAllQuery() {
        return "select * from concerts";
    }

    @Override
    protected String getSaveQuery() {
        return "insert into concerts (artist, date, location, available_tickets) values (?, ?, ?, ?)";
    }

    @Override
    protected void setSaveParameters(PreparedStatement statement, Concert entity) throws SQLException {
        statement.setString(1, entity.getArtist());
        statement.setTimestamp(2, java.sql.Timestamp.valueOf(entity.getDate()));
        statement.setString(3, entity.getLocation());
        statement.setInt(4, entity.getAvailableTickets());
    }

    @Override
    protected String getDeleteQuery() {
        return "delete from concerts where id = ?";
    }

    @Override
    protected String getUpdateQuery() {
        return "update concerts set artist = ?, date = ?, location = ?, available_tickets = ? where id = ?";
    }

    @Override
    protected void setUpdateParameters(PreparedStatement statement, Concert entity) throws SQLException {
        statement.setString(1, entity.getArtist());
        statement.setTimestamp(2, java.sql.Timestamp.valueOf(entity.getDate()));
        statement.setString(3, entity.getLocation());
        statement.setInt(4, entity.getAvailableTickets());
        statement.setLong(5, entity.getId());
    }

    @Override
    public List<Concert> findAllByDate(LocalDateTime date) {
        logger.info("Se incearca conexiunea la baza de date");
        List<Concert> concerts = new ArrayList<>();
        try (Connection connection = DriverManager.getConnection(url, username, password);
             PreparedStatement statement = connection.prepareStatement("select * from concerts where cast(date as DATE) = ?")
        ) {
            statement.setDate(1, Date.valueOf(date.toLocalDate()));
            ResultSet resultSet = statement.executeQuery();
            while (resultSet.next()) {
                Concert concert = extractEntity(resultSet);
                logger.info("Entitatea a fost gasita: " + concert);
                concerts.add(concert);
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return concerts;
    }
}

File: MPP/csharp/csharp/repository/ConcertDatabaseRepositoryImpl.cs
using System.Data;
using csharp.domain;
using Npgsql;
using Serilog;

namespace csharp.repository;

public class ConcertDatabaseRepositoryImpl : AbstractDatabaseRepository<long, Concert>, IConcertRepository
{
    public ConcertDatabaseRepositoryImpl(string? url) : base(url)
    {
    }

    protected override Concert ExtractEntity(NpgsqlDataReader resultSet)
    {
        return new Concert(
            resultSet.GetInt64("id"),
            resultSet.GetString("artist"),
            resultSet.GetDateTime("date"),
            resultSet.GetString("location"),
            resultSet.GetInt32("available_tickets")
        );
    }

    protected override string GetFindOneQuery()
    {
        return "select * from concerts where id = @id";
    }

    protected override void SetIdParameter(NpgsqlCommand statement, long id)
    {
        statement.Parameters.AddWithValue("@id", id);
    }


    protected override string GetFindAllQuery()
    {
        return "select * from concerts";
    }


    protected override string GetSaveQuery()
    {
        return "insert into concerts (id, artist, date, location, available_tickets) values (@id, @artist, @date, @location, @available_tickets)";
    }

    protected override void SetSaveQueryParameters(NpgsqlCommand statement, Concert entity)
    {
        statement.Parameters.AddWithValue("@artist", entity.Artist);
        statement.Parameters.AddWithValue("@date", entity.Date);
        statement.Parameters.AddWithValue("@location", entity.Location);
        statement.Parameters.AddWithValue("@available_tickets", entity.AvailableTickets);
    }

    protected override string GetDeleteQuery()
    {
        return "delete from concerts where id = @id";
    }

    protected override string GetUpdateQuery()
    {
        return "update concerts set artist = @artist, date = @date, location = @location, available_tickets = @available_tickets where id = @id";
    }

    protected override void SetUpdateQueryParameters(NpgsqlCommand statement, Concert entity)
    {
        statement.Parameters.AddWithValue("@id", entity.Id);
        statement.Parameters.AddWithValue("@artist", entity.Artist);
        statement.Parameters.AddWithValue("@date", entity.Date);
        statement.Parameters.AddWithValue("@location", entity.Location);
        statement.Parameters.AddWithValue("@available_tickets", entity.AvailableTickets);
    }

    public List<Concert> FindAllByDate(DateTime dateTime)
    {
        List<Concert> concerts = new List<Concert>();
        Log.Information("Se incearca conexiunea cu baza de date");
        using (NpgsqlConnection connection = new NpgsqlConnection(url))
        {
            connection.Open();
            using (NpgsqlCommand statement = connection.CreateCommand())
            {
                statement.CommandText = "select * from concerts where cast(date as DATE) = @date";
                statement.Parameters.AddWithValue("@date", dateTime.Date);
                using (NpgsqlDataReader resultSet = statement.ExecuteReader())
                {
                    while (resultSet.Read())
                    {
                        Concert concert = ExtractEntity(resultSet);
                        Log.Information("S-a gasit entitatea cu id-ul: " + concert.Id);
                        concerts.Add(concert);
                    }
                }
            }
        }
        return concerts;
    }
}